packages/protocol/contracts-0.8/common/UsingPrecompiles.sol

```diff
@@ -73,59 +74,103 @@ contract UsingPrecompiles is IsL2Check {
+      require(success, "error calling validatorSignerAddressFromCurrentSet precompile");
+      return address(uint160(getUint256FromBytes(out, 0)));
+    }
   }
 
   /**
-   * @notice Gets a validator address from the validator set at the given block number.
+   * @notice Gets a validator address from the current validator set.
    * @param index Index of requested validator in the validator set.
-   * @param blockNumber Block number to retrieve the validator set from.
    * @return Address of validator at the requested index.
    */
+  function validatorAddressFromCurrentSet(uint256 index) public view onlyL2 returns (address) {
+    address[] memory _elected = getEpochManager().getElected();
+    return _elected[index];
+  }
+
+  /**
+   * @notice Gets a validator signer address from the validator set at the given block number.
+   * @param index Index of requested validator in the validator set.
+   * @param blockNumber Block number to retrieve the validator set from.
+   * @return Address of validator signer at the requested index.
+   */
   function validatorSignerAddressFromSet(
     uint256 index,
     uint256 blockNumber
   ) public view returns (address) {
-    bytes memory out;
-    bool success;
-    (success, out) = GET_VALIDATOR.staticcall(abi.encodePacked(index, blockNumber));
-    require(success, "error calling validatorSignerAddressFromSet precompile");
-    return address(uint160(getUint256FromBytes(out, 0)));
+    if (isL2()) {
+      return getAccounts().getValidatorSigner(validatorAddressFromSet(index, blockNumber));
+    } else {
+      bytes memory out;
+      bool success;
+      (success, out) = GET_VALIDATOR.staticcall(abi.encodePacked(index, blockNumber));
+      require(success, "error calling validatorSignerAddressFromSet precompile");
+      return address(uint160(getUint256FromBytes(out, 0)));
+    }
+  }
+
+  /**
+   * @notice Gets a validator address from the validator set at the given block number.
+   * @param index Index of requested validator in the validator set.
+   * @param blockNumber Block number to retrieve the validator set from.
+   * @return Address of validator at the requested index.
+   */
+  function validatorAddressFromSet(
+    uint256 index,
+    uint256 blockNumber
+  ) public view onlyL2 returns (address) {
+    (, , , , address[] memory _elected) = getEpochManager().getEpochByBlockNumber(blockNumber);
+    return _elected[index];
   }
 
   /**
    * @notice Gets the size of the current elected validator set.
    * @return Size of the current elected validator set.
    */
   function numberValidatorsInCurrentSet() public view virtual returns (uint256) {
-    bytes memory out;
-    bool success;
-    (success, out) = NUMBER_VALIDATORS.staticcall(abi.encodePacked(uint256(block.number)));
-    require(success, "error calling numberValidatorsInCurrentSet precompile");
-    return getUint256FromBytes(out, 0);
+    if (isL2()) {
+      return getEpochManager().getElected().length;
```

Comment:
Ditto, implementing this in EpochManager should save gas since the whole `elected` array won't be read into memory.

Reply:
The limitation in implementing this whole function in `EpochManager`, is to avoid importing the `UsingPrecompiles` contract.

I created an `EpochManager` function that returns the number instead of getting the elected array and getting its length.